### PR TITLE
perf(gatsby-source-contentful): prevent redundant fs/remote fetches for tracedSVG

### LINKED ISSUE #28903: [gatsby-source-contentful] local image processing can break if the same image is requested multiple times
## Description

While processing Contentful images using the tracedSVG query, the build breaks with this kind of errors:
```
VipsJpeg: Premature end of JPEG file
VipsJpeg: Premature end of JPEG file
VipsJpeg: out of order read at line 0
VipsJpeg: out of order read at line 0
```

### Steps to reproduce

It's a little difficult to reproduce because it's not deterministic, but in theory you should simply use the same image in Contentful and query it in multiple places. For example use it in different entries, like in different pages, then use tracedSVG inside a query.

### Expected result

The build should complete without errors.

### Actual result

The build breaks with this errors:
```
VipsJpeg: Premature end of JPEG file
VipsJpeg: Premature end of JPEG file
VipsJpeg: out of order read at line 0
VipsJpeg: out of order read at line 0
```

### Environment

System:
    OS: Windows 10 10.0.18363
    CPU: (8) x64 Intel(R) Core(TM) i5-8265U CPU @ 1.60GHz
  Binaries:
    Node: 12.16.1 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.4 - ~\AppData\Roaming\npm\yarn.CMD
    npm: 6.13.4 - C:\Program Files\nodejs\npm.CMD
  Languages:
    Python: 2.7.8
  Browsers:
    Chrome: 87.0.4280.88
    Edge: Spartan (44.18362.449.0)
  npmPackages:
    gatsby: ^2.29.3 => 2.29.3
    gatsby-image: ^2.8.0 => 2.8.0
    gatsby-plugin-google-tagmanager: ^2.8.0 => 2.8.0
    gatsby-plugin-gtag: ^1.0.13 => 1.0.13
    gatsby-plugin-image: ^0.4.1 => 0.4.1
    gatsby-plugin-manifest: ^2.9.1 => 2.9.1
    gatsby-plugin-postcss: ^2.3.13 => 2.3.13
    gatsby-plugin-preact: ^4.4.0 => 4.4.0
    gatsby-plugin-react-helmet: ^3.7.0 => 3.7.0
    gatsby-plugin-react-helmet-canonical-urls: ^1.4.0 => 1.4.0
    gatsby-plugin-react-svg: ^3.0.0 => 3.0.0
    gatsby-plugin-robots-txt: ^1.5.5 => 1.5.5
    gatsby-plugin-sass: ^2.8.0 => 2.8.0
    gatsby-plugin-schema-snapshot: ^1.4.0 => 1.4.0
    gatsby-plugin-sharp: ^2.11.2 => 2.11.2
    gatsby-plugin-sitemap: ^2.9.0 => 2.9.0
    gatsby-plugin-webpack-bundle-analyser-v2: ^1.1.18 => 1.1.18
    gatsby-source-contentful: ^4.3.1 => 4.3.1
    gatsby-source-filesystem: ^2.8.1 => 2.8.1
    gatsby-transformer-json: ^2.8.0 => 2.8.0
    gatsby-transformer-remark: ^2.13.1 => 2.13.1
    gatsby-transformer-sharp: ^2.9.0 => 2.9.0
    gatsby-transformer-yaml: ^2.8.0 => 2.8.0
  npmGlobalPackages:
    gatsby-cli: 2.12.115

## Notes on troubleshooting

I've already investigated the problem. First of all, the jpeg images I used are valid.

Based on my search, the root of the error is in **gatsby-source-contentful**, specifically in the **cache-image.js** file: if the file already exists, returns readily the path for sharp to process it.
https://github.com/gatsbyjs/gatsby/blob/ef12dca08b960ca4c09a59d3e687ec7bd20e7e55/packages/gatsby-source-contentful/src/cache-image.js#L48
The problem is, if the same image is requested more than once in parallel, the second request tries to process the image while it's still downloading it, resulting in invalid image errors.
I think it could be fixed by adding a promise cache, so that every subsequent request just awaits the same promise until the file is completely downloaded.
Something like this:
```javascript
const flyingRequests = {};
....
const alreadyExists = await pathExists(absolutePath);

if (!alreadyExists && !flyingRequests[absolutePath]) {
    flyingRequests[absolutePath] = new Promise(async (resolve, reject) => {
      const previewUrl = `http:${url}?${params.join(`&`)}`;
      const response = await axios({
        method: `get`,
        url: previewUrl,
        responseType: `stream`
      });
      const file = createWriteStream(absolutePath);
      response.data.pipe(file);
      file.on(`finish`, resolve);
      file.on(`error`, reject);
    });
  }
  if (flyingRequests[absolutePath]) {
    await flyingRequests[absolutePath];
    delete flyingRequests[absolutePath];
  }

  return absolutePath;
```
I've tried this on my machine and it seems to prevent the bug. 
If you agree, I can open a pull request with a fix.

### PR DESCRIPTION
**Pull request by ariadne-github**
## Description

This adds a promise cache to prevent incoming requests of the same image from access the file on disk before it's completely downloaded. Like before, on subsequent requests after the file is fully downloaded the path is returned immediately. This includes new builds.

## Related Issues

Fixes #28903.

### CODE REVIEW

File: packages/gatsby-source-contentful/src/cache-image.js [hunk 1]
```diff
@@ -45,24 +47,36 @@ module.exports = async function cacheImage(store, image, options) {
...
-
   if (!alreadyExists) {
-    const previewUrl = `http:${url}?${params.join(`&`)}`
-
-    const response = await axios({
-      method: `get`,
-      url: previewUrl,
-      responseType: `stream`,
-    })
+    const downloadPromise = await new Promise((resolve, reject) => {
```
Comment (pieh) [inline packages/gatsby-source-contentful/src/cache-image.js:60]: There is `await` used in promise executor so we need to add `async` to function 
```suggestion
    const downloadPromise = await new Promise(async (resolve, reject) => {
```

**Comment by pieh** [comment]
I pushed some ordering change - i swapped order of file exist check and in-flight lookup check - with us checking in-flight cache before async `fileExists` - it was still causing multiple concurrent downloads because it was working like this:

tested scenario: we have to make 2 calls for image A
`call 1` - call `cacheImage`, check in-flight cache (empty), start checking fs (`pathExists`) and while async I/o happens another `cacheImage` can start executing
`call 2` - call `cacheImage`, check in-flight cache (still empty, because `pathExists` for first call didn't resolve yet) -> start checking fs

with current change it seems to work now ok - let me know what are your thoughts on my change

**Comment by ariadne-github** [comment]
I hadn't noticed that the file existence check was async, I think you're right about that. But the inflight check must be checked also before, otherwise you theoretically could fall into the original bug:

- call 1: check for the file, does not find it, check for the promise, does not find it, starts downloading
- call 2: check for the file, finds it, returns, but the file is still being downloaded

I committed a new change that should both prevent concurrent network calls, and early returns before the file is completely downloaded, which is the original and worse problem, since it breaks builds.
What do you think?

**Comment by pieh** [comment]
Yeah, you are right, I didn't even notice that we we could have writing in progress state, your reording (again 😅 ) change make sense!

**Review by pieh — APPROVED** [review]
Thanks!

**Pull request merged**